packages/components/src/components/assay/AssayProvider.tsx

```diff
@@ -0,0 +1,89 @@
+import React from 'react';
+
+import { WithRouterProps } from 'react-router';
+
+import { Alert, AssayDefinitionModel, AssayProtocolModel, getActionErrorMessage, LoadingPage, NotFound } from '../..';
+
+import { AssayStateModel } from './models';
+
+interface IContext {
+    assayDefinition: AssayDefinitionModel;
+    assayProtocol: AssayProtocolModel;
+}
+
+interface AssayLoadProps {
+    loadAssay: (protocolName: string) => any;
+    assay: AssayStateModel;
+}
+
+export interface AssayProviderProps extends WithRouterProps {
+    assayDefinition: AssayDefinitionModel;
+    assayProtocol: AssayProtocolModel;
+}
+
+type Props = AssayProviderProps & AssayLoadProps;
+
+interface State extends IContext {}
+
+const Context = React.createContext<IContext>(undefined);
+const AssayContextProvider = Context.Provider;
+export const AssayContextConsumer = Context.Consumer;
+
+export const AssayProvider = (Component: React.ComponentType) => {
+    return class AssayProviderImpl extends React.Component<Props, State> {
+        constructor(props: Props) {
+            super(props);
+
+            this.state = {
+                assayDefinition: undefined,
+                assayProtocol: undefined,
+            };
+        }
+
+        componentDidMount() {
+            this.props.loadAssay(this.props.params.protocol);
+        }
+
+        componentWillReceiveProps(nextProps: Props) {
+            if (this.props.params.protocol !== nextProps.params.protocol) {
+                this.props.loadAssay(nextProps.params.protocol);
+            }
+        }
+
+        render() {
+            const { assay, params } = this.props;
+            const { protocol } = params;
+            const assayDefinition = assay.getByName(protocol);
+
+            if (assay.isLoaded && assayDefinition) {
+                if (!assay.getProtocol(assayDefinition.id)) {
+                    return <LoadingPage />;
+                }
+
+                const state = {
```

Comment:
I think this is a great idea to use `React.Context` here to provide assay modeling.

With that, creating a new object with each `render()` will cause all context consumers to re-render every time. This component should be updating it's own `state` when the definition is loaded and then always handing `this.state` to `<AssayContextProvider value={this.state}>`. See any of our other usages of `React.createContext` within `@labkey/components` for examples.

My suggestion would be to use `getDerivedStateFromProps` and derive the new state once the assay definition and protocol are loaded.

**Edit:** I now see this was copied from SM's implementation but my feedback still holds.